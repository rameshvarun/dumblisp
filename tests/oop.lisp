;; Inspired by Scheming With Objects
(defun NEW-POINT (x y)
  ;; Getters
  (defun get-x () x)
  (defun get-y () y)
  
  ;; Setters
  (defun set-x (newx)
    (set x newx))
  (defun set-y (newy)
    (set y newy))

  ;; Printer
  (defun disp ()
    (print "(" (self 'get-x)
      "," (self 'get-y) ")"))
  
  ;; Dispatcher function
  (defun self args
    (set message (head args))
    (set args (tail args))
    (case message
      ('get-x (apply get-x args))
      ('get-y (apply get-y args))
      ('set-x (apply set-x args))
      ('set-y (apply set-y args))
      ('disp (apply disp args))
    ))

  ;; Return dispatcher
  self)

(set a (NEW-POINT 0 0))
(a 'disp)
(a 'set-x 10)
(a 'set-y 20)
(a 'disp)

(set b (NEW-POINT 0 0))
(b 'disp)
(b 'set-x 102)
(b 'set-y 20123)

(a 'disp)
(b 'disp)