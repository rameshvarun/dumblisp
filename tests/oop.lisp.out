( 0 , 0 )
( 10 , 20 )
( 0 , 0 )
( 10 , 20 )
( 102 , 20123 )
